[Vengeance] implement Void Reaver, Frailty, Soulcrush

ToppleTheNun opened the pull request:
dependent on #7064 or a replacement for it, up to you, @EvanMichaels.

**Before**
![beforesoulcrush](https://user-images.githubusercontent.com/1672786/206367027-e287cf02-8f93-4c4c-9ea8-069db4033484.png)
**After**
![image](https://user-images.githubusercontent.com/1672786/206367065-574d59e8-ccc2-4c64-bb09-0b16e069b008.png)
**Bonus: 5 target cleave**
![image](https://user-images.githubusercontent.com/1672786/206367097-24df0eb6-e72d-4385-b78e-34d4a728852e.png)

ToppleTheNun committed:
[Vengeance] initial pass at passive talents

ToppleTheNun committed:
[Vengeance] implement Void Reaver, Frailty, Soulcrush

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -5889,8 +5992,9 @@ void demon_hunter_t::init_spells()
-  
+
   spec.fiery_brand_dr = talent.vengeance.fiery_brand->ok() ? find_spell( 207744 ) : spell_data_t::not_found();
+  spec.frailty_debuff = talent.vengeance.frailty->ok() ? find_spell( 247456 ) : spell_data_t::not_found();
```
You may need the debuff to be found without frailty talented as isn't it possible to proc it while skipping that talent technically via other talents?

ToppleTheNun replied:
If you tried to build a tree without it, it would be the 8th point: https://www.wowhead.com/talent-calc/demon-hunter/vengeance/DAAAERFBUBA

EvanMichaels replied:
Got it, sounds good.

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -6654,9 +6758,9 @@ void demon_hunter_t::target_mitigation( school_e school, result_amount_type dt,
     }
 
-    if ( td->debuffs.void_reaver )
+    if ( td->debuffs.frailty->up() && talent.vengeance.void_reaver )
```
This should probably be check() rather than up() due to not wanting to have another source of benefit tracking. (Think it uses up() in the damage reduction calculations.)

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -1463,6 +1480,11 @@ class demon_hunter_action_t : public Base
 
+    if ( affected_by.frailty )
+    {
+      m *= 1.0 + p()->spec.frailty_debuff->effectN( 4 ).percent() * td( target )->debuffs.frailty->stack();
```
Should use check() here rather than stack()

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -1342,6 +1343,19 @@ class demon_hunter_action_t : public Base
 
+    ab::apply_affecting_aura( p->talent.vengeance.perfectly_balanced_glaive );
+    ab::apply_affecting_aura( p->talent.vengeance.meteoric_strikes );
+    ab::apply_affecting_aura( p->talent.vengeance.shear_fury );
```
Unfortunately since this is done as Spell Effect 2 indirection rather than a mod aura, you'll have to add this manually into the Shear energize_amount in the shear_t() constructor.

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -1342,6 +1343,19 @@ class demon_hunter_action_t : public Base
+    ab::apply_affecting_aura( p->talent.vengeance.perfectly_balanced_glaive );
+    ab::apply_affecting_aura( p->talent.vengeance.meteoric_strikes );
+    ab::apply_affecting_aura( p->talent.vengeance.shear_fury );
+    ab::apply_affecting_aura( p->talent.vengeance.roaring_fire );
```
Roaring Fire will have to be done manually since it's conditional.

ToppleTheNun replied:
Even though it's setup as an `APPLY_AURA` on the spell data?

EvanMichaels replied:
```
 #1 (id=1028507)  : Apply Aura (6) | Modify Healing% Based on Target Health% (354)
                    Base Value: 50 | Scaled Value: 50 | PvP Coefficient: 1.00000 | Misc Value: 2 | Target: Self (1)
                    Affected Spells: Fel Devastation (212106), Fel Devastation (333120)
```

This isn't an aura type we support manually applying. Since it would dynamically scale based on your health. Think I would implement it as an action_multiplier() override on fel_devastation_heal_t probably.

ToppleTheNun replied:
I'll pull that out into a separate PR, good to know!

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -1342,6 +1343,19 @@ class demon_hunter_action_t : public Base
+    ab::apply_affecting_aura( p->talent.vengeance.meteoric_strikes );
+    ab::apply_affecting_aura( p->talent.vengeance.shear_fury );
+    ab::apply_affecting_aura( p->talent.vengeance.roaring_fire );
+    ab::apply_affecting_aura( p->talent.vengeance.agonizing_flames );
```
This is already hooked up in the Immolation Aura buff.

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -1342,6 +1343,19 @@ class demon_hunter_action_t : public Base
+    ab::apply_affecting_aura( p->talent.vengeance.shear_fury );
+    ab::apply_affecting_aura( p->talent.vengeance.roaring_fire );
+    ab::apply_affecting_aura( p->talent.vengeance.agonizing_flames );
+    ab::apply_affecting_aura( p->talent.vengeance.extended_spikes );
```
Will need to be added to the Demon Spikes buff rather than the ability.

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -1342,6 +1343,19 @@ class demon_hunter_action_t : public Base
+    ab::apply_affecting_aura( p->talent.vengeance.extended_spikes );
+    ab::apply_affecting_aura( p->talent.vengeance.burning_blood );
+    ab::apply_affecting_aura( p->talent.vengeance.ruinous_bulwark );
+    ab::apply_affecting_aura( p->talent.vengeance.volatile_flameblood );
```
Volatile Flameblood will need to be implemented manually as a gain in the Immolation Aura impact (is it per hit? or per execute?)

Look at Isolated Prey handling in Fel Rush as an example.

ToppleTheNun replied:
I'll pull that out into a separate PR then

EvanMichaels reviewed (changes requested):
Added comments for everything that stood out to me. Hope that's helpful!

ToppleTheNun committed:
update from Koji's feedback

- use check() instead of up() where applicable
- use check() instead of stack() where applicable
- remove conditional apply_affecting_aura usage
- remove apply_affecting_aura that need to go in new PR

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -6654,9 +6758,9 @@ void demon_hunter_t::target_mitigation( school_e school, result_amount_type dt,
     }
 
-    if ( td->debuffs.void_reaver )
+    if ( td->debuffs.frailty->check() && talent.vengeance.void_reaver )
```
talent.vengeance.void_reaver->ok()

EvanMichaels commented on engine/class_modules/sc_demon_hunter.cpp:
```diff
@@ -6654,9 +6758,9 @@ void demon_hunter_t::target_mitigation( school_e school, result_amount_type dt,
+    if ( td->debuffs.frailty->check() && talent.vengeance.void_reaver )
     {
-      s->result_amount *= 1.0 + td->debuffs.void_reaver->stack_value();
+      s->result_amount *= 1.0 + spec.frailty_debuff->effectN( 3 ).percent() * td->debuffs.frailty->stack();
```
This one can also be
> ... * td->debuffs.frailty->check()
as stack() has benefit tracking as well

ToppleTheNun committed:
update from Koji's feedback

- ensure we check "ok()" on talents
- remove another source unnecessary benefit tracking
